### PR TITLE
212 - Document LOQ, Review and add Workflow Diagrams

### PR DESCRIPTION
**Pull request by andreievg**
Fixes: https://github.com/openmsupply/application-manager-server/issues/212

Original issue was for document review and consolidation extras, I've changed it to be about documenting LOQ and Review (this seemed like a good place to start, since I wasn't happy about lack of documentation for those parts, sorry it's long time coming).

I've added a few new diagrams, and used some existing diagrams (relocated them to new drive folder IRIMS/Internal Documentation Diagrams/Web App)

Quite a lot of documentation also relates to server, but I didn't want to abstract it too much and get lost in context spreading between different documentation sets (maybe there is a better way to organise this).

Also I've added overall workflow diagrams to `Application And Review Flow`. I haven't reviewed this PR myself for typos yet, will try to do it tomorrow.

Btw, changes (including diagrams) can be easily viewed in VS code (can preview MD, and it shows diagrams)

### CODE REVIEW

File: documentation/internal/Level-1-Review-And-LOQ.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+In the context of this document, we are assuming that reviewer is already assigned, all sections and they are the only level reviewer defined for a given stage.
+
+## Brief Summary
+
+Below is the summary of logic aspects of this topic
+
+### Back End
+
+- Changes status of review and application to inform front end of actions that can be made
```
Comment (nmadruga) [inline documentation/internal/Level-1-Review-And-LOQ.md:19]:
```suggestion
- Changes `review.status` and `application.status` to inform front end of actions that can be made
```

File: documentation/internal/Level-1-Review-And-LOQ.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+- Trims Responses that are duplicated and not changed
```
Comment (nmadruga) [inline documentation/internal/Level-1-Review-And-LOQ.md:20]:
```suggestion
- On review **submission** trim `review_responses` that have been created but not reviewed by user
- On application **re-submission**, trim `application_response` that are duplicated and not changed
- On review **re-submission**, trim `review_responses` that are duplicated and not changed
```

File: documentation/internal/Level-1-Review-And-LOQ.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+### Front End
+
+- Calculates actions and progress of review and application
+- GraphQL mutation to create review and application responses
+- GraphQL mutation to update review_responses
```
Comment (nmadruga) [inline documentation/internal/Level-1-Review-And-LOQ.md:26]:
```suggestion
- GraphQL mutation to update `application_response` or `review_response`
```

File: documentation/internal/Level-1-Review-And-LOQ.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+- Validates submission of review and application
+- GraphQL mutation to request submission of review and application
+
+### Structure
+
+The 'state' of application and review (actions, progress, etc..) is calculated and stored in one state structure. It aids visual component and de-couples logic from UI elements. This structure is recalculated whenever a response is changed
```
Comment (nmadruga) [inline documentation/internal/Level-1-Review-And-LOQ.md:32]: That's very good, I just missed a little more explanation around what hooks runs in what case, would we require a little more around that? For example, does the `useGetApplicationStructure` re-runs when there is an update to `review_response`?

File: documentation/internal/Level-1-Review-And-LOQ.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+## Schema
+
+The aim of this diagram is to give some context (an anchor point), it's an abstracted schema diagram, not all of the tables involved in Review and LOQ process are shown
+
+![Review and LOQ Schema](images/Review-And-LOQ-Schema.png)
+
+This schema structure allows new responses to be created (previous responses to be duplicated) for every new update, and this creates a system which maintains historic data integrity.
```
Comment (nmadruga) [inline documentation/internal/Level-1-Review-And-LOQ.md:40]:
```suggestion
This schema structure allows new responses to be created (previous responses to be duplicated) for every change on the application status, and this creates a system which maintains historic data integrity.
```

File: documentation/internal/Level-1-Review-And-LOQ.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+## Starting and Re-Staring Review
+
+**Front End**
+
+When reviewer start/re-starts a review, new review_responses are created for every `latest` application_response. If latest application_response already has a review_response linked (from previous review), review_response is duplicated. `useRestartReview` and `useCreateReview` hooks.
```
Comment (nmadruga) [inline documentation/internal/Level-1-Review-And-LOQ.md:46]:
```suggestion
When reviewer start/re-starts a review, new `review_response` entities are created for every `latest` `application_response`. If latest `application_response` already has a `review_response` linked (from previous review), `review_response` is duplicated. Done by `useRestartReview` and `useCreateReview` hooks.
```

File: documentation/internal/Level-1-Review-And-LOQ.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+`Latest` application response is just application response with latest time_updated for a given template_element.
```
Comment (nmadruga) [inline documentation/internal/Level-1-Review-And-LOQ.md:48]:
```suggestion
`latest_application_response` is just application response with latest `time_updated` for a given element.
```

File: documentation/internal/Level-1-Review-And-LOQ.md [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+It's possible that latest application response is not visible (it could become invisible due to changes in application), these are ignored.
+
+We only create review response for template_elements that are assigned to reviewer via review_question_assignment (see [Assignment Process](Assignment-Process.md) for more details)
```
Comment (nmadruga) [inline documentation/internal/Level-1-Review-And-LOQ.md:52]:
```suggestion
We only create `review_response` for elements that are assigned to reviewer via `review_question_assignment` (see [Assignment Process](Assignment-Process.md) for more details)
```

File: documentation/internal/Level-1-Review-And-LOQ.md [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+**Back End**
+
+When review is submitted, an action will `trim` any responses that haven't changed and will update latest review responses time_updated to match review submission time (latest from status and stage history table) `trimResponses` action.
```
Comment (nmadruga) [inline documentation/internal/Level-1-Review-And-LOQ.md:56]:
```suggestion
When review is submitted, an action will `trim` any responses that haven't changed and will update latest review responses `time_updated` to match review submission time (latest from status and stage history table) `trimResponses` action.
```

File: documentation/internal/Level-1-Review-And-LOQ.md [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+We will be restricting editing of review tables (apart from `trigger`) via postgres row level policies. Front end requests status change by setting `trigger` on review table to `onReviewCreate`, `onReviewSubmit`, `onReviewRestart`, and back end changes this via `changeStatus` action. Front end needs to wait for `trigger` to reset to `null` before displaying new state of review
+
+## Re-Staring Application
+
+In response to LOQ, an applicant can restart an application. Similar logic to the one described in above section applies.
+
+## Review Decision
+
+Is created for each iteration of a review, can be of status: `'List of questions`, `Conform`, `Non-conform`, `No Decision`. See `Validation of Submission` below for choices that will be available to the user.
```
Comment (nmadruga) [inline documentation/internal/Level-1-Review-And-LOQ.md:68]:
```suggestion
Is created for each iteration of a review, can be of `status`: 
- `List of questions`
- `Conform`
- `Non-conform`
- Default: `No Decision`

 See `Validation of Submission` below for choices that will be available to the user.
```

File: documentation/internal/Level-1-Review-And-LOQ.md [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+## Review Responses Visibility
+
+`is_visible_to_applicant` field (not editable by any user, only by back end) controls visibility of review_response from applicant perspective. And `recommended_applicant_visibility` field (which is editable by reviewer) suggests visibility of review_response to applicant. `updateReviewVisibility` back end action will set `is_visible_to_applicant` when review is submitted. See `LOQ Workflow Diagram` below
```
Comment (nmadruga) [inline documentation/internal/Level-1-Review-And-LOQ.md:72]:
```suggestion
`is_visible_to_applicant` field (not editable by any user, only by back end) controls visibility of `review_response` from applicant perspective. And `recommended_applicant_visibility` field (which is editable by reviewer) suggests visibility of review_response to applicant. `updateReviewVisibility` back end action will set `is_visible_to_applicant` when review is submitted. See `LOQ Workflow Diagram` below
```

File: documentation/internal/Level-1-Review-And-LOQ.md [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+Actions for applications `changeRequestsProgress` and `isChangeRequest` is used to display actions in Application Home pages, those are generated in `useGetApplicationStructure` -> `addApplicantChangeRequestStatusToElement` and `generateApplicantChangesRequestedProgress`.
+
+See `Structure` below for more details
+
+## Validation of Submission
+
+When **application** is restarted as a result of LOQ, it can only be submitted if general validation passes AND all questions with change requests have been changed. Logic for this is:
+
+latestApplicationResponse is different to previousApplicationResponse where previousApplicationResponse is linked to non confirm reviewResponse
```
Comment (nmadruga) [inline documentation/internal/Level-1-Review-And-LOQ.md:90]:
```suggestion
`latestApplicationResponse` is different to `previousApplicationResponse` where `previousApplicationResponse` is linked to non confirm `reviewResponse`
```

File: documentation/internal/Level-1-Review-And-LOQ.md [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+![Applicant LOQ Structure Changes](images/Applicant-LOQ-Structure-Changes.png)
+
+### Review
+
+![Level-1-Review-Structure-Changes](images/Level-1-Review-Structure-Changes.png)
+
+## Components and Hooks
+
+This is described in Application and Review Flow documentation
```
Comment (nmadruga) [inline documentation/internal/Level-1-Review-And-LOQ.md:110]: Link?

File: documentation/internal/Level-1-Review-And-LOQ.md [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+# LOQ Workflow Diagram
+
+![LOQ Workflow](images/LOQ-Workflow.png)
```
Comment (nmadruga) [inline documentation/internal/Level-1-Review-And-LOQ.md:114]: Is this related to `recommend_applicant-visibility`?
Screen Shot 2021-04-22 at 12 22 43 PM

Also, the bold is half-way through in the first `Visibility field` part of this diagram:
Screen Shot 2021-04-22 at 12 23 56 PM

File: documentation/internal/Level-1-Review-And-LOQ.md [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+# Responses Workflow Diagram
+
+Note this diagram may be slightly outdated, but it is still indicative of implemented logic. It also goes beyond Review and LOQ topic
+
+![Responses Workflow](images/Responses-Workflow.png)
```
Comment (nmadruga) [inline documentation/internal/Level-1-Review-And-LOQ.md:120]: Haven't done a review for this one. Will leave it for re-review, but have in mind it doesn't need to be exact up to date, just reflect what we have been doing currently.

**Review by nmadruga — CHANGES_REQUESTED** [review]
Many suggestions! Hopefully this will be helpful, I thought it was good to have a deeper look, better than just going quickly over all this, since there is quite a lot included in this apparently simple PR : )
 
I'll finish my review here - haven't done the Review part of the `Application-and-Review-Workflow`, but will get back to it on my re-review of this PR.

Very good to have this, thanks for putting an extra effort on getting this done. You have documented the steps to well in this diagrams that will be great having this also in our wiki!

**Comment by nmadruga** [comment]
@CarlosNZ & @andreievg Also suggest merging this one.

Have a question for Andrei as well: Have you started to make a documentation PR to list what was added for latest changes on Consolidation? I have just created an issue to do that, if you like I can work on it tomorrow, unless you have already started. Will be great to have that with existing diagrams so we can look and update during development of [Epic #78 - Review make changes requested](536)

File: documentation/internal/Application-And-Review-Flow.md [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+
+TODO add written description
+
+## Overall Diagram
+
+![Application and Review Workflow.png](images/Application-and-Review-Workflow.png)
+
+## Review Decision Diagram
+
+![Review Decision Workflow](images/Review-Decision-Workflow.png)
```
[the remark this answers is not in the document]
Comment (andreievg) [reply documentation/internal/Application-And-Review-Flow.md:11]: > The non conform? rectangle maybe should be updated to some review response set as non conform? 

I think it's meant to indicate that `non coform` is coming from canSubmitReviewAs part of review structure (could be more conditions then just review response set as non conform, like for change request).

> Should be renamed to useGetReviewStructureForSection: 

84ff2eb8ee018cc4755689bcc6c614904e65d1a7

File: documentation/internal/Application-And-Review-Flow.md [hunk 17]
```diff
@@ -0,0 +1,15 @@
+# Application and Review Flow
+
+TODO add written description
```
[the remark this answers is not in the document]
Comment (andreievg) [reply documentation/internal/Application-And-Review-Flow.md:3]: https://github.com/openmsupply/application-manager-web-app/issues/729

File: documentation/internal/Application-And-Review-Flow.md [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+
+## Overall Diagram
+
+![Application and Review Workflow.png](images/Application-and-Review-Workflow.png)
+
```
[the remark this answers is not in the document]
Comment (andreievg) [reply documentation/internal/Application-And-Review-Flow.md:8]: > useGetReviewStructure renamed to useGetReviewStructureForSection in here:

84ff2eb8ee018cc4755689bcc6c614904e65d1a7

> Also not sure if it should mention the Async version and what is it used for?

https://github.com/openmsupply/application-manager-web-app/issues/729

> Tiny one: route should be application/:id here:

84ff2eb8ee018cc4755689bcc6c614904e65d1a7

> Regarding ApplicationSummaryPage (the file is called ApplicationSummary, but I like ApplicationSummaryPage, nicer to update the file on the repo):

84ff2eb8ee018cc4755689bcc6c614904e65d1a7

> It will redirect to the page with the problem, we agree to not display the Summary with problems:

84ff2eb8ee018cc4755689bcc6c614904e65d1a7

> This one is also done, right? We do trim responses not filled or not changes (on changes_required):

Removed the bit about visibility and added a little bit extra about ON_APPLICATION_SUBMIT

84ff2eb8ee018cc4755689bcc6c614904e65d1a7

> Regarding ApplicationPage, this seems to be a little lost in there : )

I just removed for now, should be done in this issue: https://github.com/openmsupply/application-manager-web-app/issues/729

> Regarding NextButton, it is actually inside component Navigation. Also good to be affirmative at this point so the docs can answer questions we (or others) might have during development:
> Also included in the Navigation component:

Added an issue for this diagram to reflect exactly what's been implemented: https://github.com/openmsupply/application-manager-web-app/issues/729

> Renamed to PageElements and receives as props also isSummary (and instead of isEditable, canEdit)

84ff2eb8ee018cc4755689bcc6c614904e65d1a7

> Perhaps this should have a special docs to described the use of this and SectionWrapper for many pages in the the app : )

Added as part of the same issue above: https://github.com/openmsupply/application-manager-web-app/issues/729

> While you are working on this, might be good to also rename useGetApplicationFullStructure with useGetApplicationStructure which is the latest version of this hook (not important though):
> Maybe this comment can be removed, I think this isn't happening anymore:

84ff2eb8ee018cc4755689bcc6c614904e65d1a7: 

> Couldn't find this one on the code, is it still used? Maybe needs to be renamed (is to the left of useGetApplicationFullStructure block : )):

This is not up to date anymore, revalidation workflow describes it better, added as part of next doc issue, there is a small comment in the diagram that says, For Revalidation See (Revalidation Workflow.drawio), i wonder if that's enough: https://github.com/openmsupply/application-manager-web-app/issues/729
 
> Maybe like it was done for ApplicationWrapper describe main route: /application/:id/review:
>And this should be just the same as the root, we aren't using home:
> Not needed - can be removed (path to it as well):

84ff2eb8ee018cc4755689bcc6c614904e65d1a7

File: documentation/internal/Level-1-Review-And-LOQ.md (hunk 4 above)
Comment (andreievg) [reply documentation/internal/Level-1-Review-And-LOQ.md]: I've added a little bit more: d887cf212f7301f79931fee9c9a53db6409fbc3b, didn't quite add about review_response only affecting only useGetReviewStructureForSection only.  Was hoping that diagram would describe it better:

Screen Shot 2021-05-21 at 12 22 08 AM

File: documentation/internal/Level-1-Review-And-LOQ.md (hunk 13 above)
Comment (andreievg) [reply documentation/internal/Level-1-Review-And-LOQ.md]: 911facfd91e4148c0054bdaf5e8d4a5e50d38966

File: documentation/internal/Level-1-Review-And-LOQ.md (hunk 14 above)
Comment (andreievg) [reply documentation/internal/Level-1-Review-And-LOQ.md:122]: > Is this related to recommend_applicant-visibility?
yep: 6c760856c386cb9425471d64fbdca35dd048eb99
> Also, the bold is half-way through in the first Visibility field part of this diagram:
6c760856c386cb9425471d64fbdca35dd048eb99

File: documentation/internal/Level-1-Review-And-LOQ.md (hunk 15 above)
Comment (andreievg) [reply documentation/internal/Level-1-Review-And-LOQ.md:128]: I think it's generally pretty accurate. It should give a good idea of the implementation. I wish there was an easier way to update diagram and even link them to code they relate to

**Comment by andreievg** [comment]
@nmadruga 

> Have a question for Andrei as well: Have you started to make a documentation PR to list what was added for latest changes on Consolidation? I have just created an issue to do that, if you like I can work on it tomorrow, unless you have already started. Will be great to have that with existing diagrams so we can look and update during development of Epic #78 - Review make changes requested

I would recommend documenting consolidation once we've finished full functionality (including third level consolidation), as there could be some changes we haven't thought about

**Comment by andreievg** [comment]
@nmadruga thank you for the thorough review, sorry it's taken this long to make the changes.

I've addressed every comment with a commit or link to this issue: https://github.com/openmsupply/application-manager-web-app/issues/729. Also committed all but one of your helpful suggestions.

File: documentation/internal/Application-And-Review-Flow.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (nmadruga) [reply documentation/internal/Application-And-Review-Flow.md:8]: 👍 

I have added issue #729 as part of [Epic 78](https://github.com/openmsupply/application-manager-web-app/issues/536) so we can work on it as soon as possible - although it seems a bit everywhere. Hopefully everything can be addressed at the same time.

File: documentation/internal/Application-And-Review-Flow.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (nmadruga) [reply documentation/internal/Application-And-Review-Flow.md:11]: Got it, I wasn't understanding exactly that this was just meant to say how to show decision options - Was thinking it was the logic on how to set available options to the review

> > The non conform? rectangle maybe should be updated to some review response set as non conform?
> 
> I think it's meant to indicate that `non coform` is coming from canSubmitReviewAs part of review structure (could be more conditions then just review response set as non conform, like for change request).

**Review by nmadruga — APPROVED** [review]
LGTM
